sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+        s = s[mask]
+
+        u_t = u.t()
+        v_t = v.t()
+        s = torch.diag(s)
+
+        vamp_score = torch.trace(
+            2. * torch.chain_matmul(s, u_t, c0t, v)
+            - torch.chain_matmul(s, u_t, c00, u, s, v_t, ctt, v)
+        )
+    else:
+        raise RuntimeError("This should have been caught earlier.")
+    return 1 + vamp_score
+
+
+def loss(data: torch.Tensor, data_lagged: torch.Tensor, method='VAMP2', epsilon: float = 1e-6, mode: str = 'trunc'):
+    r"""Loss function that can be used to train VAMPNets. It evaluates as :math:`1-\mathrm{score}`. The score
+    is implemented in :meth:`score`."""
+    return -1. * score(data, data_lagged, method=method, epsilon=epsilon, mode=mode)
+
+
+class MLPLobe(nn.Module):
+    r""" A multilayer perceptron which can be used as a neural network lobe for VAMPNets. """
+
+    def __init__(self, units: List[int], nonlinearity=nn.ELU, initial_batchnorm: bool = True,
+                 output_nonlinearity=lambda: nn.Softmax(dim=1)):
+        r"""Instantiates a new lobe.
+
+        Parameters
+        ----------
+        units : list of integers
+            The units of the fully connected layers.
+        nonlinearity : callable, default=torch.nn.ELU
+            A callable (like a constructor) which yields an instance of a particular activation function.
+        initial_batchnorm : bool, default=True
+            Whether to use batch normalization before the data enters the rest of the network.
+        output_nonlinearity : callable, default=softmax
+            The output activation/nonlinearity. If the data decomposes into states, it can make sense to use
+            an output activation like softmax which produces a probability distribution over said states.
+        """
+        super().__init__()
+        layers = []
+        if initial_batchnorm:
+            layers.append(nn.BatchNorm1d(units[0]))
+        for fan_in, fan_out in zip(units[:-2], units[1:-1]):
+            layers.append(nn.Linear(fan_in, fan_out))
+            layers.append(nonlinearity())
+        layers.append(nn.Linear(units[-2], units[-1]))
+        if output_nonlinearity is not None:
+            layers.append(output_nonlinearity())
+        self._sequential = nn.Sequential(*layers)
+
+    def forward(self, inputs):
+        return self._sequential(inputs)
+
+
+class VAMPNetModel(Model, Transformer):
+    r"""
+    A VAMPNet model which can be fit to data optimizing for one of the implemented VAMP scores.
+
+    See Also
+    --------
+    VAMPNet : The corresponding estimator.
+    """
+
+    def __init__(self, lobe: nn.Module, lobe_timelagged: Optional[nn.Module] = None,
+                 dtype=np.float32, device=None):
+        r"""Creates a new VAMPNet estimator instance.
+
+        Parameters
+        ----------
+        lobe : torch.nn.Module
+            One of the lobes of the VAMPNet. See also :class:`MLPLobe`.
+        lobe_timelagged : torch.nn.Module, optional, default=None
```

Comment:
does lobe sharing mean sharing structure only or also weights

Reply:
weights

Reply:
clarify in docs?

Reply:
done